sw/device/lib/dif/dif_clkmgr.h

```diff
@@ -0,0 +1,200 @@
+  kDifClkmgrToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDifClkmgrToggleDisabled,
+} dif_clkmgr_toggle_t;
+
+/**
+ * Hardware instantiation parameters for clock manager.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
+ */
+typedef struct dif_clkmgr_params {
+  /**
+   * The base address for the clock manager hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_clkmgr_params_t;
+
+/**
+ * A handle to clock manager.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_clkmgr {
+  dif_clkmgr_params_t params;
+} dif_clkmgr_t;
+
+/**
+ * The result of a clock manager operation.
+ */
+typedef enum dif_clkmgr_result {
+  /**
+   * Indicates that the operation succeeded.
+   */
+  kDifClkmgrOk = 0,
+  /**
+   * Indicates some unspecified failure.
+   */
+  kDifClkmgrError = 1,
+  /**
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
+   *
+   * When this value is returned, no hardware operations occured.
+   */
+  kDifClkmgrBadArg = 2,
+} dif_clkmgr_result_t;
+
+/**
+ * An Index of a "Gateable" Clock.
+ *
+ * "Gateable" clocks are under full software control: they can be enabled and
+ * disabled by software, which directly starts and stops the identified clock.
+ */
+typedef uint32_t dif_clkmgr_gateable_clock_t;
+
+/**
+ * An Index of a "Hintable" Clock.
+ *
+ * "Hintable" clocks are under partial software control: software can suggest
+ * they are gated, but the clock manager may delay gating the peripheral, or may
+ * ignore the request altogether.
+ */
+typedef uint32_t dif_clkmgr_hintable_clock_t;
+
+/**
+ * Creates a new handle for clock manager.
+ *
+ * This function does not actuate the hardware.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param[out] handle Out param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_clkmgr_result_t dif_clkmgr_init(dif_clkmgr_params_t params,
+                                    dif_clkmgr_t *handle);
+
+/**
+ * Check if a Gateable Clock is Enabled or Disabled.
+ *
+ * @param handle Clock Manager Handle.
+ * @param clock Gateable Clock ID.
+ * @param[out] is_enabled whether the clock is enabled or not.
+ * @returns The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_clkmgr_result_t dif_clkmgr_gateable_clock_is_enabled(
+    const dif_clkmgr_t *handle, dif_clkmgr_gateable_clock_t clock,
+    bool *is_enabled);
+
+/**
+ * Enable or Disable a Gateable Clock.
+ *
+ * @param handle Clock Manager Handle.
+ * @param clock Gateable Clock ID.
+ * @param new_state whether to enable or disable the clock.
+ * @returns The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_clkmgr_result_t dif_clkmgr_gateable_clock_toggle(
+    const dif_clkmgr_t *handle, dif_clkmgr_gateable_clock_t clock,
+    dif_clkmgr_toggle_t new_state);
```

Comment:
What do you think about
```
dif_clkmgr_gateable_get_enabled
dif_clkmgr_gateable_set_enabled
```
in analogy with the IRQ functions? (Same for hintable clocks; I dropped the `_clock` path component but that's just me.)